mm2src/kdf_walletconnect/src/connection_handler.rs

```diff
@@ -0,0 +1,132 @@
+        }
+    }
+
+    fn inbound_error(&mut self, error: ClientError) {
+        debug!("[{}] inbound error: {error}", self.name);
+        if let Err(e) = self.conn_live_sender.start_send(Some(error.to_string())) {
+            error!("[{}] failed to send to the receiver: {e}", self.name);
+        }
+    }
+
+    fn outbound_error(&mut self, error: ClientError) {
+        debug!("[{}] outbound error: {error}", self.name);
+        if let Err(e) = self.conn_live_sender.start_send(Some(error.to_string())) {
+            error!("[{}] failed to send to the receiver: {e}", self.name);
+        }
+    }
+}
+
+/// Establishes initial connection to WalletConnect relay server with linear retry mechanism.
+/// Uses increasing delay between retry attempts starting from INITIAL_RETRY_SECS.
+/// After successful connection, attempts to restore previous session state from storage.
+pub(crate) async fn initialize_connection(wc: Arc<WalletConnectCtx>) {
+    info!("Initializing WalletConnect connection");
+    let mut retry_count = 0;
+    let mut retry_secs = INITIAL_RETRY_SECS;
+
+    while let Err(err) = wc.connect_client().await {
+        retry_count += 1;
+        error!(
+            "Error during initial connection attempt {}: {:?}. Retrying in {retry_secs} seconds...",
+            retry_count, err
+        );
+        Timer::sleep(retry_secs).await;
+        retry_secs += RETRY_INCREMENT;
+    }
+
+    // Initialize storage
+    if let Err(err) = wc.session.storage().init().await {
+        error!("Unable to initialize WalletConnect persistent storage: {err:?}. Only inmemory storage will be utilized for this Session.");
+    };
+
+    // load session from storage
+    if let Err(err) = wc.load_session_from_storage().await {
+        error!("Unable to load session from storage: {err:?}");
+    };
+
+    // Spawn session disconnection watcher.
+    handle_disconnections(&wc).await;
+}
+
+/// Handles unexpected disconnections from WalletConnect relay server.
+/// Implements exponential backoff retry mechanism for reconnection attempts.
+/// After successful reconnection, resubscribes to previous topics to restore full functionality.
+pub(crate) async fn handle_disconnections(this: &WalletConnectCtx) {
+    let mut recv = this.connection_live_rx.lock().await;
+    let mut backoff = 1;
+
+    while let Some(msg) = recv.next().await {
+        info!("WalletConnect disconnected with message: {msg:?}. Attempting to reconnect...");
+
+        loop {
+            match this.reconnect_and_subscribe().await {
+                Ok(_) => {
+                    info!("Reconnection process complete.");
+                    backoff = 1;
+                    break;
+                },
+                Err(e) => {
+                    info!("Reconnection attempt failed: {:?}. Retrying in {:?}...", e, backoff);
+                    Timer::sleep(backoff as f64).await;
+                    backoff = std::cmp::min(backoff * 2, MAX_BACKOFF);
```

Comment:
`60s` looks to be the perfect backoff time imo, in some cases if user gets connected just few secs after hitting 1 hour backoff then we'll have unnecessary long delay. 

I also think using back off is better than sleeping every `60secs`. We want to connect as quick as possible

Reply:
it just feels weird to me because i see the backoff strategy usually used with longer wait times. but 60s with backoff vs 60s constant isn't much of a difference so im cool with any :).